core/drivers/clk/fixed_clk.c

```diff
@@ -0,0 +1,76 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2021, Bootlin
+ */
+
+#include <drivers/clk.h>
+#include <drivers/clk_dt.h>
+#include <libfdt.h>
+#include <malloc.h>
+#include <stdint.h>
+
+struct fixed_clock_data {
+	unsigned long rate;
+};
+
+static unsigned long fixed_clk_get_rate(struct clk *clk,
+					unsigned long parent_rate __unused)
+{
+	struct fixed_clock_data *d = clk->priv;
+
+	return d->rate;
+}
+
+static const struct clk_ops fixed_clk_clk_ops = {
+	.get_rate = fixed_clk_get_rate,
+};
+
+static TEE_Result fixed_clock_setup(const void *fdt, int offs)
+{
+	const uint32_t *freq = NULL;
+	const char *name = NULL;
+	struct clk *clk = NULL;
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct fixed_clock_data *fcd = NULL;
+
+	name = fdt_get_name(fdt, offs, NULL);
+	if (!name)
+		name = "fixed-clock";
+
+	clk = clk_alloc(name, &fixed_clk_clk_ops, NULL, 0);
+	if (!clk)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	fcd = calloc(1, sizeof(struct fixed_clock_data));
+	if (!fcd) {
+		res = TEE_ERROR_OUT_OF_MEMORY;
+		goto free_clk;
+	}
+
+	freq = fdt_getprop(fdt, offs, "clock-frequency", NULL);
+	if (!freq) {
+		res = TEE_ERROR_BAD_FORMAT;
+		goto free_fcd;
+	}
+
+	fcd->rate = fdt32_to_cpu(*freq);
+	clk->priv = fcd;
+
+	res = clk_register(clk);
+	if (res)
+		goto free_fcd;
+
+	res = clk_dt_register_clk_provider(fdt, offs, clk_dt_get_simple_clk,
+					   clk);
+	if (!res)
+		return 0;
```

Comment:
`return TEE_SUCCESS`

minor: I would suggest the following:
```c
	res = clk_dt_register_clk_provider(fdt, offs, clk_dt_get_simple_clk,
					   clk);
out:
	if (res) {
		free(fcd);
		clk_free(clk);
	}

	return res;
```

Reply:
Wasn't sure of that for the end check. In some way, your suggestion duplicates the existing code for freeing but my `(!res)` is less intuitive.